File: reservoir.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include "reservoir.h"
#include "math_utils.h"

struct Reservoir* create_reservoir(
    size_t num_neurons, size_t num_inputs, size_t num_outputs, 
    double learning_rate, double spectral_radius, double ei_ratio, double input_strength, double connectivity, double dt,
    enum ConnectivityType connectivity_type, enum NeuronType neuron_type, double *neuron_params) {

    struct Reservoir *reservoir = malloc(sizeof(*reservoir));
    if (reservoir == NULL) {
        fprintf(stderr, "Error allocating memory for reservoir of size %zu\n", num_neurons);
        return NULL;
    }
    
    reservoir->num_neurons = num_neurons;
    reservoir->num_inputs = num_inputs;
    reservoir->num_outputs = num_outputs;
    reservoir->learning_rate = learning_rate;
    reservoir->spectral_radius = spectral_radius;
    reservoir->ei_ratio = ei_ratio;
    reservoir->input_strength = input_strength;
    reservoir->connectivity = connectivity; 
    reservoir->dt = dt;
    reservoir->connectivity_type = connectivity_type;
    reservoir->neuron_type = neuron_type;
    
    reservoir->neurons = (void**)malloc(num_neurons * sizeof(void*));
    
    if (!(reservoir->neurons)) {
        fprintf(stderr, "Memory allocation failed for reservoir neurons\n");
        free(reservoir); 
        return NULL;
    }

    for (size_t i = 0; i < num_neurons; i++) {
        reservoir->neurons[i] = init_neuron(neuron_type, neuron_params); 
    }

    return reservoir;
}


// todo break inputs off to send only to 'num_inputs' neurons


/* old version
void step_reservoir(struct Reservoir *r, double input) {
    for (size_t i = 0; i < r->num_neurons; i++) {
        double neuron_input = input * r->W_in[i] * r->input_strength;

        for (size_t j = 0; j < r->num_neurons; j++) {
            neuron_input += get_neuron_spike(r->neurons[j], r->neuron_type) * r->W[i * r->num_neurons + j];
        }

        update_neuron(r->neurons[i], r->neuron_type, neuron_input, r->dt);
    }
}
*/


void step_reservoir(struct Reservoir *r, double input) {
    if (!r || r->dt <= 0) return;

    // The macro step duration is implicitly 1.0.
    // Calculate how many internal micro-steps to run.
    int num_micro_steps = (int)(1.0 / r->dt);
    size_t num_neurons = r->num_neurons;

    // --- Pre-calculate constant external inputs for this entire macro step ---
    double external_inputs[num_neurons];
    for (size_t i = 0; i < num_neurons; i++) {
        external_inputs[i] = input * r->W_in[i] * r->input_strength;
    }

    // --- Buffer to hold spikes, which evolve at the micro-step scale ---
    // We get the spikes from the previous macro_step to start.
    double last_spikes[num_neurons];
    for (size_t i = 0; i < num_neurons; i++) {
        last_spikes[i] = get_neuron_spike(r->neurons[i], r->neuron_type);
    }

    // --- Internal simulation loop (the "micro-steps") ---
    for (int t = 0; t < num_micro_steps; t++) {
        for (size_t i = 0; i < num_neurons; i++) {
            // 1. Calculate recurrent input based on spikes from the LAST micro-step
            double recurrent_input = 0.0;
            for (size_t j = 0; j < num_neurons; j++) {
                recurrent_input += last_spikes[j] * r->W[i * num_neurons + j];
            }

            // 2. Total input is the sum of the constant external and dynamic recurrent inputs
            double total_input = external_inputs[i] + recurrent_input;

            // 3. Update the neuron's internal state by one 'dt'
            update_neuron(r->neurons[i], r->neuron_type, total_input, r->dt);
        }

        // 4. Record all spikes that occurred in THIS micro-step to be used in the NEXT one
        for (size_t i = 0; i < num_neurons; i++) {
            last_spikes[i] = get_neuron_spike(r->neurons[i], r->neuron_type);
        }
    }
}

/****** OLD VERSION ******/
/*
void step_reservoir(struct Reservoir *reservoir, double input) {
    for (size_t i = 0; i < reservoir->num_neurons; i++) {
        double neuron_input = input * reservoir->W_in[i] * reservoir->input_strength;
        for (size_t j = 0; j < reservoir->num_neurons; j++) {
            // add inputs from other spiking neurons
            neuron_input += get_neuron_spike(reservoir->neurons[j], reservoir->neuron_type) * reservoir->W[i * reservoir->num_neurons + j];
        } 
        update_neuron(reservoir->neurons[i], reservoir->neuron_type, neuron_input);
    }
}

void run_reservoir(struct Reservoir *reservoir, double *inputs, size_t input_length) {
    for (size_t i = 0; i < input_length; i++) {
        double input = inputs[i]; 
        step_reservoir(reservoir, input); 
    }
}
*/


// compute_output calculates the sum of neuron states * W_out
double compute_output(struct Reservoir *reservoir) {
    double output = 0.0;
    for (size_t i = 0; i < reservoir->num_neurons; i++) {
        output += get_neuron_state(reservoir->neurons[i], reservoir->neuron_type) * reservoir->W_out[i];
    }

    return output;
}

// compute_activity returns the sum of spikes
double compute_activity(struct Reservoir *reservoir) {
    double total_activity = 0.0;
    double spike; 
    for (size_t i = 0; i < reservoir->num_neurons; i++) {
        spike = get_neuron_spike(reservoir->neurons[i], reservoir->neuron_type);
        total_activity += spike;
    }
    
    return total_activity;
}

// return the neuron->V for each neuron
double *read_reservoir_state(struct Reservoir *reservoir) {
    double *state = malloc(reservoir->num_neurons * sizeof(double));
    if (state == NULL) {
        fprintf(stderr, "Memory allocation error for reservoir state\n");
        return NULL;
    }
    for (size_t i = 0; i < reservoir->num_neurons; i++) {
        state[i] = get_neuron_state(reservoir->neurons[i], reservoir->neuron_type);
    }
    return state; // caller will need to free()
}

void train_output_iteratively(struct Reservoir *reservoir, double target) {
    double *state = read_reservoir_state(reservoir); // malloc'd in function, free here
    if (state == NULL) { return; }
    double prediction = compute_output(reservoir);
    double error = target - prediction;

    // update W_out
    for (size_t i =0; i < reservoir->num_neurons; i++) {
        reservoir->W_out[i] += state[i] * error * reservoir->learning_rate;
    }

    free(state);
}

void free_reservoir(struct Reservoir *reservoir) {
    if (!reservoir) { return; }
    
    printf("Freeing reservoir at address: %p\n", (void*)reservoir);
    printf("Freeing neurons at address: %p\n", (void*)reservoir->neurons);
    printf("Freeing W_in at address: %p\n", (void*)reservoir->W_in);
    printf("Freeing W_out at address: %p\n", (void*)reservoir->W_out);
    printf("Freeing W at address: %p\n", (void*)reservoir->W);

    for (size_t i = 0; i < reservoir->num_neurons; i++) {
        free_neuron(reservoir->neurons[i], reservoir->neuron_type);
        reservoir->neurons[i] = NULL;
    }
    free(reservoir->neurons);
    free(reservoir->W_in);
    free(reservoir->W_out);
    free(reservoir->W);
    reservoir->W_in = NULL;
    reservoir->W_out = NULL;
    reservoir->W = NULL;
    free(reservoir);
    reservoir = NULL;
}


double generate_weight(double ei_ratio) {
    double magnitude = (double)rand() / RAND_MAX;
    if (((double)rand() / RAND_MAX) < ei_ratio) {
        return magnitude;
    }
    else {
        return -magnitude;
    }
}

/**
 * @brief Initializes the weight matrices for a reservoir.
 * * @param reservoir A pointer to the reservoir to initialize.
 * @return EXIT_SUCCESS or EXIT_FAILURE.
 * * @note The user is responsible for seeding the random number generator 
 * before initializing a reservoir. This should be done once at the beginning
 * of the main() application by calling srand(time(NULL)).
 */

int init_weights(struct Reservoir *reservoir) {
    reservoir->W_in = malloc(reservoir->num_neurons * sizeof(double)); 
    if(reservoir->W_in == NULL) {
        fprintf(stderr, "Error allocating memory for W_in, size of reservoir: %zu\n", reservoir->num_neurons);
        return EXIT_FAILURE;
    }
    
    reservoir->W_out = calloc(reservoir->num_neurons, sizeof(double)); // set output weights to zero initially
    if(reservoir->W_out == NULL) {
        fprintf(stderr, "Error allocating memory for W_out, size of reservoir: %zu\n", reservoir->num_neurons);
        free(reservoir->W_in);
        reservoir->W_in = NULL; 
        return EXIT_FAILURE;
    }
    
    // calloc here to ensure non-assigned weights are automatically clamped to 0.0 to simplify the matrix operations
    reservoir->W = calloc(reservoir->num_neurons * reservoir->num_neurons, sizeof(double)); 
    if (reservoir->W == NULL) {
        fprintf(stderr, "Error allocating memory for W, size of reservoir: %zu\n", reservoir->num_neurons);
        free(reservoir->W_in);
        free(reservoir->W_out);
        reservoir->W_in = NULL;
        reservoir->W_out = NULL;
        return EXIT_FAILURE;
    }

    switch(reservoir->connectivity_type) {
        case RANDOM:
            for (size_t i = 0; i < reservoir->num_neurons; i++) {
                reservoir->W_in[i] = generate_weight(reservoir->ei_ratio);
                for(size_t j = 0; j < reservoir->num_neurons; j++) {
                    if (i == j) continue; // no-self connections
                    if(((double)rand() / RAND_MAX) < reservoir->connectivity) {
                        reservoir->W[i * reservoir->num_neurons + j] = generate_weight(reservoir->ei_ratio);
                    }
                } 
            }
            break;

        case SMALL_WORLD:
            fprintf(stderr, "Small-world connectivity not yet implemented.\n");
            break;

        case SCALE_FREE:
            fprintf(stderr, "Scale-free connectivity not yet implemented.\n");
            break;
    }

    return EXIT_SUCCESS;
}

int rescale_weights(struct Reservoir *reservoir) {
    // rescale weights to ensure spectral radius
    double current_spectral_radius = calc_spectral_radius(reservoir->W, reservoir->num_neurons);
    if (current_spectral_radius > 1e-9) { // avoid division by zero
        double rescaling_factor = reservoir->spectral_radius / current_spectral_radius; 
        for (size_t i = 0; i < (reservoir->num_neurons * reservoir->num_neurons); i++) {
                reservoir->W[i] *= rescaling_factor;
        }
    }
    return EXIT_SUCCESS;
}

int randomize_output_layer(struct Reservoir *reservoir) {
    if (reservoir == NULL) {
        fprintf(stderr, "Failed to randomize output weights. Reservoir not initialized!");
        return EXIT_FAILURE;
    }

    double ei_ratio = 0.5; // good overall starting point for rich outputs
    for (size_t i = 0; i < reservoir->num_neurons; i++) {
        reservoir->W_out[i] = generate_weight(ei_ratio);
    }
    return EXIT_SUCCESS;
}

File: reservoir.go
package spires

import (
	"fmt"
	"math"
	"math/rand"
)

// Internal reservoir structure
type reservoir struct {
	neurons           []neuron
	numNeurons        int
	numInputs         int
	numOutputs        int
	spectralRadius    float64
	eiRatio           float64
	inputStrength     float64
	connectivity      float64
	dt                float64
	connectivityType  ConnectivityType
	neuronType        NeuronType
	neuronParams      []float64
	WIn               []float64
	WOut              []float64
	W                 []float64
}

// neuron interface for different neuron types
type neuron interface {
	update(input float64, dt float64)
	getState() float64
	getSpike() float64
	reset()
}

// createReservoir creates a new reservoir instance
func createReservoir(numNeurons, numInputs, numOutputs int,
	spectralRadius, eiRatio, inputStrength, connectivity, dt float64,
	connectivityType ConnectivityType, neuronType NeuronType, neuronParams []float64) (*reservoir, error) {

	r := &reservoir{
		numNeurons:       numNeurons,
		numInputs:        numInputs,
		numOutputs:       numOutputs,
		spectralRadius:   spectralRadius,
		eiRatio:          eiRatio,
		inputStrength:    inputStrength,
		connectivity:     connectivity,
		dt:               dt,
		connectivityType: connectivityType,
		neuronType:       neuronType,
		neuronParams:     make([]float64, len(neuronParams)),
		WIn:              make([]float64, numNeurons*numInputs),
		WOut:             make([]float64, numOutputs*numNeurons),
		W:                make([]float64, numNeurons*numNeurons),
		neurons:          make([]neuron, numNeurons),
	}

	// Copy neuron parameters
	copy(r.neuronParams, neuronParams)

	// Initialize neurons
	for i := 0; i < numNeurons; i++ {
		neuron, err := initNeuron(neuronType, neuronParams, dt)
		if err != nil {
			return nil, fmt.Errorf("failed to initialize neuron %d: %w", i, err)
		}
		r.neurons[i] = neuron
	}

	return r, nil
}

// initReservoir initializes the reservoir weights and connections
func initReservoir(r *reservoir) error {
	if r == nil {
		return fmt.Errorf("reservoir not created")
	}

	if err := initWeights(r); err != nil {
		return fmt.Errorf("failed to initialize weights: %w", err)
	}
	if err := rescaleWeights(r); err != nil {
		return fmt.Errorf("failed to rescale weights: %w", err)
	}
	if err := randomizeOutputLayer(r); err != nil {
		return fmt.Errorf("failed to randomize output layer: %w", err)
	}

	return nil
}

// initWeights initializes the weight matrices
func initWeights(r *reservoir) error {
	// Initialize input weights
	for i := 0; i < r.numNeurons; i++ {
		for j := 0; j < r.numInputs; j++ {
			r.WIn[i*r.numInputs+j] = (rand.Float64()*2 - 1) * r.inputStrength
		}
	}

	// Initialize recurrent weights based on connectivity type
	switch r.connectivityType {
	case ConnRandom:
		initRandomWeights(r)
	case ConnSmallWorld:
		initSmallWorldWeights(r)
	case ConnScaleFree:
		initScaleFreeWeights(r)
	default:
		return fmt.Errorf("unknown connectivity type: %d", r.connectivityType)
	}

	// Initialize output weights (will be trained)
	for i := 0; i < r.numOutputs; i++ {
		for j := 0; j < r.numNeurons; j++ {
			r.WOut[i*r.numNeurons+j] = 0.0
		}
	}

	return nil
}

// initRandomWeights creates random connections
func initRandomWeights(r *reservoir) {
	for i := 0; i < r.numNeurons; i++ {
		for j := 0; j < r.numNeurons; j++ {
			if i != j && rand.Float64() < r.connectivity {
				r.W[i*r.numNeurons+j] = (rand.Float64()*2 - 1)
			} else {
				r.W[i*r.numNeurons+j] = 0.0
			}
		}
	}
}

// initSmallWorldWeights creates small-world network connections
func initSmallWorldWeights(r *reservoir) {
	// Start with a regular ring lattice
	k := int(r.connectivity * float64(r.numNeurons) / 2)
	for i := 0; i < r.numNeurons; i++ {
		for j := 1; j <= k; j++ {
			idx1 := (i + j) % r.numNeurons
			idx2 := (i - j + r.numNeurons) % r.numNeurons
			r.W[i*r.numNeurons+idx1] = (rand.Float64()*2 - 1)
			r.W[i*r.numNeurons+idx2] = (rand.Float64()*2 - 1)
		}
	}

	// Rewire with some probability
	for i := 0; i < r.numNeurons; i++ {
		for j := 0; j < r.numNeurons; j++ {
			if r.W[i*r.numNeurons+j] != 0 && rand.Float64() < 0.1 {
				r.W[i*r.numNeurons+j] = 0.0
				newTarget := rand.Intn(r.numNeurons)
				if newTarget != i {
					r.W[i*r.numNeurons+newTarget] = (rand.Float64()*2 - 1)
				}
			}
		}
	}
}

// initScaleFreeWeights creates scale-free network connections
func initScaleFreeWeights(r *reservoir) {
	// Preferential attachment model
	degrees := make([]int, r.numNeurons)
	
	// Start with a few connected nodes
	for i := 0; i < 3; i++ {
		for j := i + 1; j < 3; j++ {
			r.W[i*r.numNeurons+j] = (rand.Float64()*2 - 1)
			r.W[j*r.numNeurons+i] = (rand.Float64()*2 - 1)
			degrees[i]++
			degrees[j]++
		}
	}

	// Add remaining nodes with preferential attachment
	for i := 3; i < r.numNeurons; i++ {
		// Connect to existing nodes with probability proportional to degree
		totalDegree := 0
		for _, d := range degrees {
			totalDegree += d
		}

		for j := 0; j < i; j++ {
			if rand.Float64() < float64(degrees[j])/float64(totalDegree)*r.connectivity {
				r.W[i*r.numNeurons+j] = (rand.Float64()*2 - 1)
				r.W[j*r.numNeurons+i] = (rand.Float64()*2 - 1)
				degrees[i]++
				degrees[j]++
			}
		}
	}
}

// rescaleWeights rescales the recurrent weights to achieve the desired spectral radius
func rescaleWeights(r *reservoir) error {
	currentRho := calcSpectralRadius(r.W, r.numNeurons)
	if currentRho > 0 {
		scale := r.spectralRadius / currentRho
		for i := 0; i < r.numNeurons*r.numNeurons; i++ {
			r.W[i] *= scale
		}
	}
	return nil
}

// randomizeOutputLayer initializes random output weights
func randomizeOutputLayer(r *reservoir) error {
	for i := 0; i < r.numOutputs*r.numNeurons; i++ {
		r.WOut[i] = (rand.Float64()*2 - 1) * 0.1
	}
	return nil
}

// stepReservoir advances the reservoir by one time step
func stepReservoir(r *reservoir, input []float64) {
	// Prepare input vector (zero if nil)
	u := make([]float64, r.numInputs)
	if input != nil {
		copy(u, input)
	}

	// Compute input contribution: W_in * u
	inputContribution := make([]float64, r.numNeurons)
	matVecMult(r.WIn, u, inputContribution, r.numNeurons, r.numInputs)

	// Compute recurrent contribution: W * x
	state := make([]float64, r.numNeurons)
	for i := 0; i < r.numNeurons; i++ {
		state[i] = r.neurons[i].getState()
	}
	recurrentContribution := make([]float64, r.numNeurons)
	matVecMult(r.W, state, recurrentContribution, r.numNeurons, r.numNeurons)

	// Update each neuron
	for i := 0; i < r.numNeurons; i++ {
		totalInput := inputContribution[i] + recurrentContribution[i]
		r.neurons[i].update(totalInput, r.dt)
	}
}

// runReservoir runs the reservoir on a series of inputs
func runReservoir(r *reservoir, inputSeries []float64, seriesLength int) []float64 {
	if r == nil || r.dt <= 0.0 {
		return nil
	}

	numInputs := r.numInputs
	numOutputs := r.numOutputs

	outputSeries := make([]float64, numOutputs*seriesLength)

	for i := 0; i < seriesLength; i++ {
		currentInput := inputSeries[i*numInputs : (i+1)*numInputs]
		stepReservoir(r, currentInput)

		currentOutput := outputSeries[i*numOutputs : (i+1)*numOutputs]
		computeOutput(r, currentOutput)
	}

	return outputSeries
}

// computeOutput computes the current output
func computeOutput(r *reservoir, out []float64) error {
	state := make([]float64, r.numNeurons)
	for i := 0; i < r.numNeurons; i++ {
		state[i] = r.neurons[i].getState()
	}

	// Compute output: W_out * state
	matVecMult(r.WOut, state, out, r.numOutputs, r.numNeurons)
	return nil
}

// readReservoirState returns the current neuron states
func readReservoirState(r *reservoir) []float64 {
	state := make([]float64, r.numNeurons)
	for i := 0; i < r.numNeurons; i++ {
		state[i] = r.neurons[i].getState()
	}
	return state
}

// trainOutputIteratively performs online training
func trainOutputIteratively(r *reservoir, targetVec []float64, lr float64) {
	// Get current output
	currentOutput := make([]float64, r.numOutputs)
	computeOutput(r, currentOutput)

	// Get current state
	state := make([]float64, r.numNeurons)
	for i := 0; i < r.numNeurons; i++ {
		state[i] = r.neurons[i].getState()
	}

	// Update output weights: W_out += lr * (target - output) * state^T
	for i := 0; i < r.numOutputs; i++ {
		error := targetVec[i] - currentOutput[i]
		for j := 0; j < r.numNeurons; j++ {
			r.WOut[i*r.numNeurons+j] += lr * error * state[j]
		}
	}
}

// trainOutputRidgeRegression performs ridge regression training
func trainOutputRidgeRegression(r *reservoir, inputSeries, targetSeries []float64, seriesLength, lambda float64) {
	// Collect states for all time steps
	states := make([][]float64, seriesLength)
	for t := 0; t < seriesLength; t++ {
		// Reset and run up to time t
		resetReservoir(r)
		for i := 0; i <= t; i++ {
			currentInput := inputSeries[i*r.numInputs : (i+1)*r.numInputs]
			stepReservoir(r, currentInput)
		}
		states[t] = readReservoirState(r)
	}

	// Solve ridge regression: W_out = (X^T * X + λI)^(-1) * X^T * Y
	// where X is the state matrix and Y is the target matrix
	solveRidgeRegression(r, states, targetSeries, seriesLength, lambda)
}

// solveRidgeRegression solves the ridge regression problem
func solveRidgeRegression(r *reservoir, states [][]float64, targets []float64, seriesLength int, lambda float64) {
	// This is a simplified implementation
	// In practice, you might want to use a more efficient solver
	
	// For now, we'll use a simple iterative approach
	// This could be optimized with proper matrix operations
	
	// Initialize W_out with small random values
	for i := 0; i < r.numOutputs*r.numNeurons; i++ {
		r.WOut[i] = (rand.Float64()*2 - 1) * 0.01
	}

	// Simple gradient descent
	lr := 0.001
	epochs := 1000
	
	for epoch := 0; epoch < epochs; epoch++ {
		for t := 0; t < seriesLength; t++ {
			// Compute current output
			currentOutput := make([]float64, r.numOutputs)
			matVecMult(r.WOut, states[t], currentOutput, r.numOutputs, r.numNeurons)
			
			// Compute error
			error := targets[t] - currentOutput[0] // Assuming single output for simplicity
			
			// Update weights
			for j := 0; j < r.numNeurons; j++ {
				r.WOut[j] += lr * (error*states[t][j] - lambda*r.WOut[j])
			}
		}
	}
}

// resetReservoir resets all neurons to their initial state
func resetReservoir(r *reservoir) {
	for i := 0; i < r.numNeurons; i++ {
		r.neurons[i].reset()
	}
}

// freeReservoir frees the reservoir resources
func freeReservoir(r *reservoir) {
	if r != nil {
		r.neurons = nil
		r.WIn = nil
		r.WOut = nil
		r.W = nil
		r.neuronParams = nil
	}
}

// matVecMult performs matrix-vector multiplication: y = A * x
func matVecMult(A, x, y []float64, rows, cols int) {
	for i := 0; i < rows; i++ {
		y[i] = 0
		for j := 0; j < cols; j++ {
			y[i] += A[i*cols+j] * x[j]
		}
	}
}

// calcSpectralRadius calculates the spectral radius of a matrix
func calcSpectralRadius(A []float64, n int) float64 {
	// This is a simplified implementation
	// In practice, you might want to use power iteration or other methods
	
	// For now, we'll use a simple approximation
	maxEigenvalue := 0.0
	for i := 0; i < n; i++ {
		rowSum := 0.0
		for j := 0; j < n; j++ {
			rowSum += math.Abs(A[i*n+j])
		}
		if rowSum > maxEigenvalue {
			maxEigenvalue = rowSum
		}
	}
	
	return maxEigenvalue
}
